src/subscribers/serial_subscriber.jl

```diff
@@ -0,0 +1,182 @@
+            sub.read_buffer[i] = read(sub.serial_port, UInt8)
+            if sub.read_buffer[i] == delim #0x00
+                return @view sub.read_buffer[max(1,i+1-msg_block_size):i]
+            end
+        end
+    end
+    # If serial port isn't open or flag byte isn't encountered in buffer return nothing
+    return nothing
+end
+
+"""
+    decode(msg)
+Uses [COBS](https://en.wikipedia.org/wiki/Consistent_Overhead_Byte_Stuffing)
+to decode message block.
+"""
+function decode(sub::SerialSubscriber, msg::AbstractVector{UInt8})
+    incoming_msg_size = length(msg)
+    incoming_msg_size == 0 && error("Empty message passed to encode!")
+    incoming_msg_size > msg_block_size && error("Can only safely encode 256 bytes at a time")
+
+    if !any(msg .== 0x00)
+        return nothing
+    end
+
+    push_ind, pop_ind = 1, 1
+    n = msg[pop_ind]
+    pop_ind += 1
+
+    c = 0
+    b = msg[pop_ind]
+    pop_ind += 1
+
+    # While we haven't encountered flag byte and haven't read more than the
+    # maximum message size
+    while b ≠ 0x00 && push_ind <= msg_block_size && pop_ind <= incoming_msg_size
+        c += 1
+        if c < n
+            sub.msg_in_buffer[push_ind] = b
+        else
+            sub.msg_in_buffer[push_ind] = 0x00
+
+            n = b
+            c = 0
+        end
+
+        b = msg[pop_ind]
+        pop_ind += 1
+        push_ind += 1
+    end
+
+    sub.msg_in_length = push_ind - 1
+
+    return view(sub.msg_in_buffer, 1:sub.msg_in_length)
+end
+
+"""
+Returns `true` if successfully read message from serial port
+"""
+function receive(sub::SerialSubscriber,
+                 proto_msg::ProtoBuf.ProtoType,
+                 write_lock::ReentrantLock,
+                 )
+    if isopen(sub)
+        sub.flags.isreceiving = true
+        encoded_msg = readuntil(sub, 0x00)
+        sub.flags.isreceiving = false
+
+        if encoded_msg !== nothing
+            decoded_msg = decode(sub, encoded_msg)
+            lock(write_lock) do
+                ProtoBuf.readproto(IOBuffer(decoded_msg), proto_msg)
+            end
+            # @info "Recieved message"
+
+            sub.flags.hasreceived = true
+            return true
+        end
+
+        return false
+    else
+        @warn "Attempting to receive a message on subscriber $(sub.name), which is closed"
+    end
+end
+
+"""
+Loops recieve(sub::Subscriber, proto_msg::ProtoBuf.ProtoType, write_lock=ReentrantLock())
+"""
+function subscribe(sub::SerialSubscriber,
+                   proto_msg::ProtoBuf.ProtoType,
+                   write_lock::ReentrantLock,
+                   )
+    @info "$(sub.name): Listening for message type: $(typeof(proto_msg)), on: $(sub.name)"
+    try
+        while isopen(sub)
+            receive(sub, proto_msg, write_lock)
+            GC.gc(false)
+            yield()
+        end
+        @warn "Shutting Down subscriber $(getname(sub)) on: $(tcpstring(sub)). Socket was closed."
```

Comment:
Maybe use an `@info` here instead, since this is theoretically expected behavior if the subscriber is closed correctly.